File: leetcode/Python/0187-m-repeated-dna-sequences.py
"""
Time: O(n)
Space: O(n)
"""
class Solution:
    def findRepeatedDnaSequences(self, s: str) -> List[str]:
        seen = set()
        repeated = set()
        for i in range(len(s) - 10 + 1):
            cur = s[i: i + 10]
            if cur in seen:
                repeated.add(cur)
            else:
                seen.add(cur)
        return list(repeated)

File: leetcode/C++/0187-m-repeated-dna-sequences.cpp
class Solution {
public:
    vector<string> findRepeatedDnaSequences(string s) {
        int L = 10, n = s.length();
        unordered_set<string> seen, res;
        for (int start = 0; start < n - L + 1; start++) {
            string tmp = s.substr(start, L);
            if (seen.find(tmp) != seen.end())
                res.insert(tmp);
            seen.insert(tmp);
        }
        return vector<string>(res.begin(), res.end());
    }
};